### DISCUSSION TITLE
New address parsing implementation using spaCy

### DISCUSSION THREAD

**Discussion by NickCrews** [NickCrews/mismo#68] [Show and tell]
Hey @lmores , I know you've been using mismo for address deduplication. I have a similar need, and I wasn't happy with the 2Gb size need of libpostal, and how difficult it is to install. So I went searching, and made https://github.com/NickCrews/spacy-address yesterday. I was amazed at how easy it was to get going, spaCy's developer experience is awesome. I incorporated this into a new `mismo.geo.spacy_tag_address()` function. It only is trained on US addresses and returns the labels used by the US Postal Service, so I'm not sure if it is directly useful to you, if you are working with international addresses.

- If this might solve your problem, please check it out and give me feedback
- If you want to adapt it to solve your problem, I can help with that. Pretty much all that is needed would be to come up with some training data. I only was working with ~1500 addresses, and still had great results (🤯), so it might actually be really tenable.
- I tried going with a slightly lower-level API of "*_tag_address()" instead of "*_parse_address()", because I want access to the lower level taggings for doing address deduplication (ie when comparing "123 E Elm St" and "123 Elm Avenue", I only really want to compare the "123" and "Elm" parts). I was thinking of stacking another level of API on top of that that returns a struct of street1, street2, etc, as the current postal_parse_address() does. Then, this made me think maybe we should split the postal_parse_address() into these two different API layers as well? I think I'm going to do this, let me know if you have comments.

**Comment by lmores** [comment]
Hi @NickCrews, thank you for letting me know about this, that's really interesting!

 * Yes, this is definitely something I need to solve my problem (my final goal is to achieve incremental clustering, i.e. each time a new address enters my system, I need to be able to associate it with the group of those that are equivalent to it, but that may be written in a different way)
 * I am dealing with Italian addresses (mostly).
 * Your idea to split the api into two levels makes a lot of sense, especially as the postal addresses of some geographical region may have quite different fields and the current shape of `PostalAddress` may not properly fit them all.
 * As you may have imagined, I work on this project just once a week. I will surely try out spacy-address but this will take some time.
 * Just so you know, some days ago GitHub suggested me this repository https://github.com/GRAAL-Research/deepparse which serves the same purpose as libpostal and spacy-address.
